Improve generation and use of ref/ projects

dougbu opened the pull request:
This PR is a fairly large cleanup of our 'release/3.0' work to at least get implementation assemblies built against ref/ assemblies. It has the following aims:
- remove workarounds to the extent possible
- where the above isn't possible, centralize workarounds (for simplicity and to remove redundancy)
- avoid overloading the compiler with an extreme list of referenced assemblies
- eliminate as much complexity as possible in the relevant infrastructure

Stretch goals include:
- limit bloat of the ref/ assemblies
- improve ref/ project generation bootstrapping (speedup, avoid unexpected errors)

dougbu committed:
Cleanup warnings
- avoid "CSC warning CS2008: No source files specified." building site extensions
- correct warnings (as errors) about `RenderToStringResult` being obsolete
  - add Microsoft.AspNetCore.SpaServices.Tests to Middleware solution

dotnet-maestro[bot] committed:
Improve use of ref/ assemblies
- compile against ref/ assemblies but do not change package metadata
  - update the metadata of implementation projects to include the ref/ assembly path
  - update `@(ReferenceAssembly)` metadata for Extensions packages, not `@(PackageReference)`
  - can be disabled using `$(CompileUsingReferenceAssemblies)` e.g. when generating ref/ projects
- include ref/ projects in source build by default
  - remove `$(ExcludeFromSourceBuild)` overrides from ref/ project files
- use latest package references and use project references even when _not_ building the targeting packs
- restore previous `@(Reference)` -> `@(PackageReference)` logic
  - add build-only Microsoft.Internal.Extensions.Refs package reference in most cases
- remove IndirectReferences.props and `@(_ExtensionInternalRefAssemblies)`; no longer needed

dougbu committed:
Regenerate ref/ projects
- pick up the latest generation changes

(unclear why Mvc.RazorPages/ref/Microsoft.AspNetCore.Mvc.RazorPages.netcoreapp3.0.cs changed but works)

dougbu committed:
Remove `@(RuntimeHostConfigurationOption)` workarounds
- deps files are unaffected by new ref/ assembly handling and test projects aren't special-cased
- also execute a test previously skipped due to deps file problems

dougbu committed:
Undo some manual ref/ project changes
- now done automatically or centrally
- remove manual `[TypeForwardedTo]` and `[InternalsVisibleTo]` attributes
  - fully qualify a type now that `using` is gone
- remove dupe `@(Compile)` items for *.Manual.cs files; included in the ref/ project files
- remove redundant `$(AllowUnsafeBlocks)` and `$(NoWarn)` settings

nits:
- rename a *.Manual.cs file that's not TFM-specific
- remove `private` members

dougbu committed:
Correct use of `@(ProjectReference)` items for reference providers
- use `@(Reference)` instead

dougbu committed:
Remove recently-added `@(Compile)` and `@(Reference)` items
- were added due to missing `[InternalsVisibleTo]` attributes in ref/ assemblies or as early workarounds
  - plus, now transitive references **Just Work:tm:**
- expose `ClosedGenericMatcher` in the usual (*.Manual.cs) way
- also undo Microsoft.Extensions.ApiDescription.Server workaround

dougbu committed:
Remove useless src/PackageArchive files
- not used outside 2.x branches

dougbu committed:
Perform smaller cleanup
- remove `$(IsTargetingPackPatching)`; use only `$(IsTargetingPackBuilding)`
- remove `$(DisableServicingFeatures)`; enable the servicing features we need
  - suppress baseline references even in servicing builds
- restore `$(AdditionalGenApiCmdOptions)`; useful when updating *.Manual.cs files

nits:
- simplify conditions using `$(HasReferenceAssembly)`
- correct spelling in comments
- shorten long lines

dougbu committed:
Improve ref/ project generation
- use ../src/**/AssemblyInfo.cs files instead of including attributes in *.Manual.cs files
  - for same reason, copy `@(InternalsVisibleto)` items from src/ to ref/ projects
- use eng/targets/CSharp.ReferenceAssembly.props instead of ref/Directory.Build.props files
- use TFM-specific *.Manual.cs files in ref/ project files instead of ref/Directory.Build.props files

optimizations and usability improvements:
- add `$(BuildMainlyReferenceProviders)` property to focus on reference providers when generating ref/ projects
- disable `$(UseReferenceAssemblyInImplementation)` to avoid using ref/ projects while generating them

nits:
- clean up whitespace and remove blank lines in ref/ project files

dougbu committed:
Fill in missing `internal` types 1 of n
- rename Microsoft.AspNetCore.Components.netstandard2.0.Manual.cs; need `RenderTreeFrame` type everywhere
- add types needed in unit and perf tests to *.Manual.cs files

dougbu committed:
Remove `private` members from ref/ *.Manual.cs files
- not useful and bloat the ref/ assemblies

dougbu committed:
Use a response file for GenAPI commands
- work around dotnet/arcade#4021 and help with additional ref assemblies
- mimic 111462e0c24f and integrate w/ other changes here

dougbu commented:
This PR is incomplete. Please see [Build.log](https://github.com/aspnet/AspNetCore/files/3877212/Build.log) (containing ~560 errors) and [Build.unique.log](https://github.com/aspnet/AspNetCore/files/3877213/Build.unique.log) (containing the ~190 unique errors).

In addition, the src\Identity\Extensions.Stores\ref\Microsoft.Extensions.Identity.Stores.Manual.cs file is empty. Just waiting to find what types we're using.

dougbu commented:
Steps to handle the `internal` members that test, sample and perf code is complaining about:
1. find the relevant *.cs file on disk
2. navigate to the containing src/ folder
3. `dotnet msbuild /t:GenerateReferenceSource /p:AdditionalGenApiCmdOptions="--all"`
4. `cp ..\ref\*.netcoreapp3.0.cs ..\ref\save.cs`
5. `cp ..\ref\*.netcoreapp3.0.cs ..\ref\{project}.Manual.cs` # if file doesn't exist yet
6. `dotnet msbuild /t:GenerateReferenceSource` # easy to forget! (don't forget)
7. if the *.Manual.cs file didn't exist, pare it down to the missing members
    - otherwise, find the missing members and copy them into the existing *.Manual.cs file
8. remove almost all `private` members and associated attributes
    - leave `__dummy` and `__dummyPrimitive` alone in most cases
      - but, remove them too when they're in a `struct` with `[StructLayout]` and other fields
    - leave `private` constructors calling a base constructor
    - remove `[AsyncStateMachine]` attributes; they almost always refer to `private` classes
9. `dotnet msbuild ..\test\` # to see what is still missing in the ref/ project you're changing

Suggest splitting up by ref/ project area to avoid both working on ProjectA because it contains missing members at opposite ends of the alphabet.

FYI I grabbed the warnings-only log using `"/flp:v=quiet;logfile=.\artifacts\log\Debug\Build.log"` on the build command line.

dougbu commented:
Most urgent thing is to test building (and executing) variants of e.g. src\SignalR\samples\ClientSample outside the AspNetCore repo i.e. with the 3.0.1 targeting pack / SDK and all `@(Reference)` items turned into `@(PackageReference)` items. If there be dragons, may need something drastic like building more of Extensions and all of EF Core against ref/ assemblies. (Yes, the second would be very strange given EF Core packages target .NET Standard. Worst case, the projects would need to be multi-targeted.) I'm of course hoping no problems arise and we can paper over the issues with a few additional direct references on Extensions packages.

I suggest changing Extensions very slightly to build Microsoft.Extensions.Logging.Testing and maybe one or two other internal packages against ref/ assemblies.

dougbu commented:
First 4 commits have already been merged -- in #17245. Visible because I changed the base branch to 'release/3.0' and GitHub hasn't noticed those changes are already there.

dougbu commented:
Changed my mind and rebased the branch. See you all in a few days.

wtgodbe committed:
Add Microsoft.AspNetCore.Routing.Manual.cs

wtgodbe commented on src/Http/Routing/src/Matching/CandidateState.cs:
```diff
@@ -50,6 +50,6 @@ internal CandidateState(Endpoint endpoint, RouteValueDictionary values, int scor
         /// <see cref="Http.Endpoint"/> and the current request.
         /// </summary>
-        public RouteValueDictionary Values { get; internal set; }
+        public RouteValueDictionary Values { get; set; }
```
@dougbu @JunTaoLuo I had to change this setter to not be internal, else `/t:GenerateReferenceSources` would undo the change to https://github.com/aspnet/AspNetCore/pull/17311/commits/49379ed50480afa0b0009960c8ecbcb7dace50a6#diff-24bb8cbda21395fb70004db888e179d5R520, causing a build break in the following test:

https://github.com/aspnet/AspNetCore/blob/e0ca7facf08f4bb978adc0c55d48f78d82ffdac8/src/Http/Routing/test/UnitTests/Matching/DefaultEndpointSelectorTest.cs#L42

Is this the right fix?

wtgodbe replied:
Specifically:

> DefaultEndpointSelectorTest.cs(42,13): error CS0200: Property or indexer 'CandidateState.Values' cannot be assigned to -- it is read only [E:\code\aspnet\aspnetcore\src\Http\Routing\test\UnitTests\Microsoft.AspNetCore.Routing.Tests.csproj]

Pilchie replied:
Tagging @rynowak.  I don't think that's the right fix.  If we can't do this another way, I would suggest instead removing the setter, and making an internal `SetValues()` method :(

wtgodbe replied:
@JunTaoLuo has done this in the past by excluding the relevant API in GenAPI, and adding the setter manually to the `manual.cs` file - I did that here: https://github.com/aspnet/AspNetCore/pull/17311/commits/b43e386df9d0968834efad3796bd9ab6f13a19a4

Pilchie replied:
Perfect - thanks.  I didn't know we had GenAPI exclusions set up to support cases like this.

JunTaoLuo replied:
Yup, I ran into similar issues in 3.0.1 and this seems like the most convenient and reliable way to resolve it.

wtgodbe committed:
Add Microsoft.AspNetCore.Authentication.AzureAD.UI.Manual.cs

wtgodbe committed:
Fixup Microsoft.AspNetCore.Routing.Manual.cs

wtgodbe committed:
Add Microsoft.AspNetCore.Authentication.AzureADB2C.UI.Manual.cs

wtgodbe commented:
I don't have anything half baked here, I've just done the manual stuff for: 

`Microsoft.AspNetCore.Routing.Manual.cs`
`Microsoft.AspNetCore.Authentication.AzureADB2C.UI.Manual.cs`
`Microsoft.AspNetCore.Authentication.AzureAD.UI.Manual.cs`

I did a bit of testing on the scenario from https://github.com/aspnet/AspNetCore/pull/17311#issuecomment-557339340 and wasn't able to repro, but I don't know if I've convinced myself that customers definitely won't be broken. A bit more testing there might be good. 

Over the weekend I'll try to generate some more `manual.cs` files - @JunTaoLuo if you do any of that, please let us know which namespaces you're working on

Commit:
Add Manual.cs files for Server projects

JunTaoLuo commented:
I updated the Server projects. I'll help with more when I get back on Tuesday.

dougbu committed:
Clean up recent commits
- remove recently-added `private` members
- restore `_dummyPrimitive` fields in Microsoft.AspNetCore.Server.HttpSys.Manual.cs

dougbu commented:
Removed and replaced `private` members as needed.

Down to 350 errors, 137 after eliminating duplicates within `{root}/src` directories, and 128 after eliminating all duplicates.

Currently working on problems in src/Hosting/&hellip;

dougbu committed:
Add *.Manual.cs files for src/Hosting projects

dougbu committed:
!fixup! fields in *.Manual.cs `struct`s
- GenAPI sometimes generates `_dummy` and `_dummyPrimitive` fields _instead of_ visible members
- what GenAPI generates sometimes have the right length but actual fields don't hurt
  - that is, using the real fields corrects both the visible API and `struct`s' sizes

nits:
- consolidate `namespace`s in Microsoft.AspNetCore.Mvc.Core.Manual.cs

dougbu committed:
Add a few more `internal` types needed in tests

dougbu commented:
Looking through src/Middleware errors next&hellip;

JunTaoLuo commented:
I'll take a look at DataProtection

Commit:
Add *.Manual.cs files for src/DataProtection projects

dougbu committed:
Add / update *.Manual.cs files for src/Middleware

dougbu committed:
Remove special case for generating ref/ projects on non-Windows
- referenced issue was closed with no action but workaround still not required
- no tabs in generated content

dougbu commented:
I updated the instructions to improve fidelity of fields in `struct`s [above](https://github.com/aspnet/AspNetCore/pull/17311#issuecomment-557337557).

Current state: I see 687 errors in a local macOS build from root. A large number of those are CS1705 issues and most of the remaining are hit in src/Mvc/Mvc.Core. [Build.log](https://github.com/aspnet/AspNetCore/files/3922999/Build.log) is the `Verbosity=quiet` log from that build. [Build.unique.per.area.log](https://github.com/aspnet/AspNetCore/files/3923000/Build.unique.per.area.log) is a unique-ified copy (excluding the CS1705 errors).

I'm going to tackle the 88 errors in src/Mvc/Mvc.Core next. Other areas with lots of errors include
- 28 errors in the rest of src/Mvc/&hellip;
- 25 errors in src/Components/&hellip;
- 3 errors in src/Security/Authentication
- 1 error in src/Identity/Extensions.Stores

Commit:
Feedback

JunTaoLuo commented on eng/targets/ResolveReferences.targets:
```diff
@@ -29,114 +30,122 @@
-        * when a project is a test or sample project
-        * the targeting pack is being patched
-      We don't use project references between components in servicing builds between compontents to preserve the baseline as much as possible.
+      Projects should use project references (instead of baseline packages), compile against ref/ assemblies, and
```
This logic is changing because we decided to ship everything all the time right?

dougbu replied:
For `$(UseProjectReferences)`, yes. The logic was wrong for `$(ReferenceReferenceAssemblies)` and `$(ReferenceImplementationAssemblies)` given everything builds against ref/ assemblies.

JunTaoLuo commented on eng/targets/ResolveReferences.targets:
```diff
@@ -265,24 +211,92 @@
+        IncludeAssets="Build"
+        PrivateAssets="All" />
+  </ItemGroup>
+  <Target Name="AddReferencePackageMetadata"
```
Clever!

JunTaoLuo commented on eng/targets/ResolveReferences.targets:
```diff
@@ -265,24 +211,92 @@
-           Code="BUILD002"
-           Text="Package references changed since the last release. This could be a breaking change and is not allowed in a servicing update. References removed:%0A - @(UnusedBaselinePackageReference, '%0A -')" />
+    <!--
+      Set properties controlling the item group GetTargetPathWithTargetPlatformMoniker, GetTargetPath and Build return.
```
Clever!

JunTaoLuo reviewed (commented):
Looked at the code in eng/targets and it looks good.

wtgodbe commented:
https://github.com/aspnet/AspNetCore/pull/17601 should be merged into this to fix the issues w/ Extensions refs

wtgodbe commented:
Still seeing errors like:

> CSC : error CS1705: Assembly 'Microsoft.AspNetCore.Blazor' with identity 'Microsoft.AspNetCore.Blazor, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60' uses 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=3.0.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' which has a higher version than referenced assembly 'Microsoft.Extensions.DependencyInjection.Abstractions' with identity 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' [F:\workspace\_work\1\s\src\Components\Blazor\Blazor\test\Microsoft.AspNetCore.Blazor.Tests.csproj]

> DbContextHealthStartup.cs(35,13): error CS1705: Assembly 'Microsoft.EntityFrameworkCore' with identity 'Microsoft.EntityFrameworkCore, Version=3.0.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' uses 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=3.0.2.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' which has a higher version than referenced assembly 'Microsoft.Extensions.DependencyInjection.Abstractions' with identity 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' [F:\workspace\_work\1\s\src\Middleware\HealthChecks\test\testassets\HealthChecksSample\HealthChecksSample.csproj]

dougbu commented:
@wtgodbe that's expected.

We've only fixed the problem for assemblies from Extensions. Could probably reduce the number of CS1705 errors by using ref/ assemblies even when targeting .NET Standard in this repo. Would then be left w/ only the errors involving EFCore assemblies. Those would be easily worked around with a limited number of direct dependencies.

I'm not positive about "using ref/ assemblies even when targeting .NET Standard" because I can imagine problems in this area since customers won't have the 3.0.0.0 assemblies when they target .NET Standard. Will experiment&hellip;

JunTaoLuo commented:
FYI I missed most of the struct members in the changes I made. I'll fixup those too. Also, currently taking a look at src/Components.

wtgodbe commented on Directory.Build.props:
```diff
@@ -84,14 +84,14 @@
+    <!-- We can remove the 3.0.1 line from any branch other than release/3.0 and from here after 3.0.1 is released. -->
+    <IsTargetingPackBuilding Condition=" '$(DotNetBuildFromSource)' == 'true' ">false</IsTargetingPackBuilding>
+    <IsTargetingPackBuilding
+        Condition=" '$(IsTargetingPackBuilding)' == '' AND '$(VersionPrefix)' == '3.0.1' ">true</IsTargetingPackBuilding>
```
We can probably just remove the "3.0.1" condition, since we're done building it

wtgodbe commented on Directory.Build.targets:
```diff
@@ -161,13 +158,12 @@
   </Target>
 
   <Import Project="eng\Workarounds.targets" />
-  <Import Project="eng\IndirectReferences.props" Condition="'$(IsTestProject)' == 'true' OR '$(IsTestAssetProject)' == 'true' OR '$(IsBenchmarkProject)' == 'true' OR '$(IsSampleProject)' == 'true'" />
```
❤️

wtgodbe commented on Directory.Build.targets:
```diff
@@ -161,13 +158,12 @@
   <Import Project="eng\targets\Wix.Common.targets"  Condition="'$(MSBuildProjectExtension)' == '.wixproj'" />
   <Import Project="eng\targets\Npm.Common.targets"  Condition="'$(MSBuildProjectExtension)' == '.npmproj'" />
-  <Import Project="eng\targets\ReferenceAssembly.targets" Condition=" '$(HasReferenceAssembly)' == 'true' " />
+  <Import Project="eng\targets\ReferenceAssembly.targets" Condition=" $(HasReferenceAssembly) " />
```
does msbuild evaluate an empty string as `false`?

dougbu replied:
`$(HasReferenceAssembly)` is no longer _ever_ empty.

wtgodbe commented on eng/CodeGen.proj:
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <BuildManaged>true</BuildManaged>
     <RepoRoot>$([MSBuild]::NormalizeDirectory('$(MSBuildThisFileDirectory)', '..'))</RepoRoot>
+    <BuildMainlyReferenceProviders>true</BuildMainlyReferenceProviders>
```
What does this property mean? Maybe add a comment

wtgodbe commented on eng/targets/ReferenceAssembly.targets:
```diff
@@ -61,37 +59,58 @@
-      <FilteredOriginalReferences
-        Include="%(_OriginalReferences.Identity)"
-        Condition="'%(_OriginalReferences.NuGetPackageId)' == '' AND '%(_OriginalReferences.PrivateAssets)' != 'All'" />
+      <_ReferenceAssemblyCompileItems Include="$(_RefSourceFileName)" />
```
It may be worth adding a comment describing what this Item captures, in case we need to add a class of files to it later

wtgodbe commented on src/Components/Blazor/testassets/Directory.Build.props:
```diff
@@ -1,10 +0,0 @@
-  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\, Directory.Build.props))\Directory.Build.props" />
-
-  <ItemGroup>
-    <!-- Workaround missing entries in deps files. Still under investigation -->
```
Why were we able to delete this file?

dougbu replied:
See eng\targets\CSharp.ReferenceAssembly.props

wtgodbe commented on src/Components/Components/ref/Directory.Build.props:
```diff
@@ -1,9 +0,0 @@
-
-  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory)..\, Directory.Build.props))\Directory.Build.props" />
-
-  <ItemGroup>
```
Why were we able to delete this file?

dougbu replied:
Because the project file was fixed to include the *.Manual.cs file.

More generally, as I mentioned in the PR's description, I got rid of numerous one-off workarounds by centralizing the logic.

wtgodbe commented on src/Security/Authentication/OAuth/ref/Microsoft.AspNetCore.Authentication.OAuth.csproj:
```diff
@@ -2,11 +2,9 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <TargetFrameworks>netcoreapp3.0</TargetFrameworks>
-    <ExcludeFromSourceBuild>false</ExcludeFromSourceBuild>
```
Is this just not necessary since `false` is the default, or do we want to start excluding this?

dougbu replied:
I changed how `$(ExcludeFromSourceBuild)` is set to avoid these numerous special cases.

wtgodbe reviewed (approved):
In general it looks good. I think it would be valuable to add a document describing how our reference resolution works now, since it's still magic to most people. Maybe add a comment in ResolveReferences.targets linking to the doc.

wtgodbe commented:
I'll start working on `src/Mvc` minus `Mvc.Core`

wtgodbe committed:
Add manual.cs for Mvc.NewtonsoftJson

dougbu commented on eng/targets/CSharp.ReferenceAssembly.props:
```diff
@@ -2,6 +2,6 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
-    <NoWarn>$(NoWarn);CS0618;PUB0001</NoWarn>
+    <NoWarn>$(NoWarn);CS0618;CS0169;PUB0001</NoWarn>
```
Where did this start showing up?

JunTaoLuo replied:
Agreed, we shouldn't add this.

wtgodbe replied:
Without this we get:

> Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.Manual.cs(433,44): error CS0169: The field 'LibuvFunctions.uv_buf_t._field1' is never used [E:\code\aspnet\aspnetcore\src\Servers\Kestrel\Transport.Libuv\ref\Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.csproj]
Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.Manual.cs(432,44): error CS0169: The field 'LibuvFunctions.uv_buf_t._field0' is never used [E:\code\aspnet\aspnetcore\src\Servers\Kestrel\Transport.Libuv\ref\Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.csproj]
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1488,21): error CS0169: The field 'IntegerDecoder._m' is never used [E:\code\aspnet\aspnetcore\src\Servers\Kestrel\Core\ref\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
Microsoft.AspNetCore.Server.Kestrel.Core.Manual.cs(1487,21): error CS0169: The field 'IntegerDecoder._i' is never used [E:\code\aspnet\aspnetcore\src\Servers\Kestrel\Core\ref\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]

It isn't clear to me how to best resolve this. It refers to:

https://github.com/aspnet/AspNetCore/blob/48e1ec445f21011086595ad080fb21174265be9d/src/Servers/Kestrel/Transport.Libuv/ref/Microsoft.AspNetCore.Server.Kestrel.Transport.Libuv.Manual.cs#L430-L435

wtgodbe replied:
Here's the struct in `src`:

https://github.com/aspnet/AspNetCore/blob/30be2bdf76d32683ad3a29d4834cb11199faecb0/src/Servers/Kestrel/Transport.Libuv/src/Internal/Networking/LibuvFunctions.cs#L429-L455

wtgodbe committed:
Add Manual file for Mvc.ApiExplorer

wtgodbe committed:
Add manual stuff for Mvc.DataAnnotations

wtgodbe committed:
Add manual stuff for Mvc.Formatters.Xml

wtgodbe committed:
Add manul file for Mvc.Localization

Commit:
Add / update *.Manual.cs files for src/Components

wtgodbe committed:
Add manual file for Authentication.Negotiate

wtgodbe committed:
Add missed reference in Extensions.Identity.Stores

wtgodbe committed:
Fixup Mvc.Viewfeatures.Manual.cs

wtgodbe committed:
fixup Mvc.NewtonsoftJson.Manual.cs

Commit:
Fixup structs in src/Components

dougbu committed:
Add more `internal` members to src/Mvc/Mvc.Core/ref

dougbu commented:
Down to 67 errors, most of which are CS1705. Remaining ones are listed in latest [Build.unique.per.area.log](https://github.com/aspnet/AspNetCore/files/3932907/Build.unique.per.area.log)

I'm going to look deeper into the CS1705 errors next&hellip;

dougbu commented:
The 17 unique non-CS1705 errors and warnings are mostly in src/Servers/Kestrel/. The remaining are in src/Mvc/Mvc.Cors, src/Mvc/Mvc.ApiExplorer, src/Mvc/Mvc.ViewFeatures and src/Identity/Extensions.Stores (one error per area). A couple of the MVC issues look weird -- as if `[InternalsVisibleTo]` isn't being honoured or a `using` is missing.

dougbu committed:
!fixup! add another `internal` type to Microsoft.AspNetCore.Mvc.Core.Manual.cs
!

dougbu commented:
Think I just fixed the "weird" couple of MVC errors.

wtgodbe commented:
I'll start on Servers\Kestrel

dougbu commented:
Thanks @wtgodbe I'm thinking the build keeps getting further; need to iterate a bit more on Mvc.Core before I tackle the CA1705 errors.

dougbu commented:
- [ ] upload test projects
- [ ] CS1705
    - [ ] extra ref/ assemblies
- [ ] src/Servers/Kestrel
- [ ] src/Mvc/Mvc.TagHelpers
- [ ] other non-CS1705 errors

dougbu committed:
!fixup! Still more `internal` Mvc.Core types

dougbu commented:
Latest [Build.log](https://github.com/aspnet/AspNetCore/files/3933567/Build.log) and [Build.unique.per.area.log](https://github.com/aspnet/AspNetCore/files/3933568/Build.unique.per.area.log)

wtgodbe committed:
Fix Servers/Kestrel with manual types

JunTaoLuo commented:
@dougbu My original repro is at https://github.com/JunTaoLuo/301Repro. It's a 3.0.1 Mvc app with a reference to EFCore 3.0.1. The conflict we were testing for is the M.E.Logging.* dependencies which is referenced as 3.0.1 by EF and 3.0.0 by Mvc. The current hypothesis is that the deps file resolves these correctly and loads 3.0.1 assemblies at runtime.

wtgodbe committed:
Add manual file for Mvc.TagHelpers

wtgodbe commented:
@dougbu looks like I lost my test projects, I had them in a side dir locally in this branch, then lost them when I checked out another branch 😢

wtgodbe commented:
In SourceBuild:

> /__w/1/s/eng/targets/ResolveReferences.targets(232,5): error MSB3245: Could not resolve this reference. Could not locate the package or project for "Microsoft.AspNetCore.Testing". Did you update baselines and dependencies lists? See docs/ReferenceResolution.md for more details. [/__w/1/s/src/Hosting/Server.IntegrationTesting/ref/Microsoft.AspNetCore.Server.IntegrationTesting.csproj]

wtgodbe commented:
I'm now looking at Mvc.ViewFeatures & Mvc.Cors

wtgodbe committed:
Add more manual types to Mvc.ViewFeatures

wtgodbe committed:
Add manual file for Mvc.Cors

wtgodbe commented:
Remaining non-CS1705 errors:

> Infrastructure/DynamicPageEndpointSelector.cs(15,26): error CS0122: 'DataSourceDependentCache<T>' is inaccessible due to its protection level [/Users/runner/runners/2.160.1/work/1/s/src/Mvc/Mvc.RazorPages/src/Microsoft.AspNetCore.Mvc.RazorPages.csproj]

>TestClient.cs(73,42): error CS0122: 'MemoryBufferWriter' is inaccessible due to its protection level [/Users/runner/runners/2.160.1/work/1/s/src/SignalR/common/testassets/Tests.Utils/Microsoft.AspNetCore.SignalR.Tests.Utils.csproj]

>TestClient.cs(77,87): error CS0122: 'MemoryBufferWriter.ToArray()' is inaccessible due to its protection level [/Users/runner/runners/2.160.1/work/1/s/src/SignalR/common/testassets/Tests.Utils/Microsoft.AspNetCore.SignalR.Tests.Utils.csproj]

>TestClient.cs(81,21): error CS0122: 'MemoryBufferWriter' is inaccessible due to its protection level [/Users/runner/runners/2.160.1/work/1/s/src/SignalR/common/testassets/Tests.Utils/Microsoft.AspNetCore.SignalR.Tests.Utils.csproj]

wtgodbe committed:
Fixup Mvc.Razorpages manual issues

wtgodbe commented:
The `MemoryBufferWriter` issues will need a fix similar to https://github.com/aspnet/AspNetCore/pull/17311/commits/a9db76d85ea816a04bcef85720a1ee76195aff78 (they're likely hiding some other missing types). I'm heading home, if no one beats me to it I'll pick that up on Monday.

dougbu commented:
Rebased to pick up a few more fixes e.g. @wtgodbe's cafb50c68b08 and 89aae45d6dce

dougbu committed:
Only create ref/ projects for assemblies in the shared framework
- restrict when `$(HasReferenceAssembly)` is `true` by default
- add warnings when `$(IsAspNetCoreApp)` or `$(HasReferenceAssembly)` have unexpected values

dougbu committed:
Remove "extra" ref/ projects
- associated implementation projects no longer have `$(HasReferenceAssembly)` set to `true`

dougbu commented:
Bit more about handling `private` members: The main reasons `__dummy`, `__dummyPrimitive` and similar `private` fields exist in what GenAPI generates are covered in https://blog.paranoidcoding.com/2016/02/15/are-private-members-api-surface.html

So, the `private` member handling I described [above](https://github.com/aspnet/AspNetCore/pull/17311#issuecomment-557337557) is correct but involves more copying than absolutely required. Because `sizeof(...)` is usually evaluated at runtime, copying all of a `struct`'s `private` fields isn't always mandatory. But, it is mandatory to check the real (implementation) type, to copy `public` and `internal` fields, and to copy everything when `[StructLayout]` is specified on the real type i.e. might as well copy the fields over.

Note: _Any_ `[StructLayout]` on the real type means the fields must be copied over w/ their offsets or (if `LayoutKind.Sequential`) in the correct order.

dougbu committed:
Add a few GenAPI exclusions
- see dotnet/arcade#4488
- generation for these members leads to NREs

dougbu committed:
Add more `internal` types and members
- Identity/Core
- Identity/Extensions.Core
- Mvc/Mvc.ViewFeatures

wtgodbe commented:
The only remaining non-`CS1705` (other than source build) is:

> F:\workspace\_work\1\s\src\Components\Web.JS\Microsoft.AspNetCore.Components.Web.JS.npmproj : error MSB4057: The target "GetTargetPath" does not exist in the project.

Plus a couple `NU5128` warnings:

> F:\workspace\_work\1\s\.dotnet\sdk\3.0.101\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: Some target frameworks declared in the dependencies group of the nuspec and the lib/ref folder do not have exact matches in the other location. Consult the list of actions below: [F:\workspace\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\Symbols\Microsoft.AspNetCore.ANCMSymbols.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\3.0.101\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: - Add lib or ref assemblies for the netcoreapp3.0 target framework [F:\workspace\_work\1\s\src\Servers\IIS\AspNetCoreModuleV2\Symbols\Microsoft.AspNetCore.ANCMSymbols.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\3.0.101\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: Some target frameworks declared in the dependencies group of the nuspec and the lib/ref folder do not have exact matches in the other location. Consult the list of actions below: [F:\workspace\_work\1\s\src\Analyzers\shared\FeatureDetection\Microsoft.AspNetCore.Analyzers.FeatureDetection.Sources.csproj]
F:\workspace\_work\1\s\.dotnet\sdk\3.0.101\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(198,5): warning NU5128: - Add lib or ref assemblies for the netstandard1.0 target framework [F:\workspace\_work\1\s\src\Analyzers\shared\FeatureDetection\Microsoft.AspNetCore.Analyzers.FeatureDetection.Sources.csproj]

Commit:
Fixup NU5128 warnings

dougbu committed:
Add direct dependencies to work around CS1705 errors
- add direct references to some test and sample projects to make intent clear i.e. address CS1705 root cause
  - these projects must use implementation assemblies for those direct references
  - requirement also applies to anything depending on them e.g. functional tests
  - for simplicity, use `$(CompileUsingReferenceAssemblies)` instead of targeted `@(Reference)` metadata
- leads to ~40 projects that do not themselves add ref/ metadata
  - this is _not_ transitive i.e. it applies only to projects that override `$(CompileUsingReferenceAssemblies)`

dougbu committed:
nits: Remove a few more `private` members in *.Manual.cs files

dougbu committed:
!fixup! correct namespaces of a few types in *.Manual.cs files

dougbu committed:
Try another way to fix Microsoft.AspNetCore.Blazor.Build.Tests

dougbu committed:
Try another way to fix missing targets in Web.JS.npmproj

dougbu committed:
Revert "Fixup NU5128 warnings"

This reverts commit 1ca248327539740eb4252623b30ff9af920223c0.
